lake/modules/strg_fifo.py

```diff
@@ -0,0 +1,475 @@
+                 data_width=16,
+                 banks=1,
+                 memory_width=64,
+                 rw_same_cycle=False,
+                 read_delay=1,
+                 addr_width=9):
+        super().__init__("strg_fifo", debug=True)
+
+        # Generation parameters
+        self.banks = banks
+        self.data_width = data_width
+        self.memory_width = memory_width
+        self.rw_same_cycle = rw_same_cycle
+        self.read_delay = read_delay
+        self.addr_width = addr_width
+        self.fw_int = int(self.memory_width / self.data_width)
+
+        assert banks > 1 or rw_same_cycle is True or self.fw_int > 1, \
+            "Can't sustain throughput with this setup. Need potential bandwidth for " + \
+            "1 write and 1 read in a cycle - try using more banks or a macro that supports 1R1W"
+
+        # Clock and Reset
+        self._clk = self.clock("clk")
+        self._rst_n = self.reset("rst_n")
+
+        # Inputs + Outputs
+        self._push = self.input("push", 1)
+        self._data_in = self.input("data_in", self.data_width)
+        self._pop = self.input("pop", 1)
+
+        self._data_out = self.output("data_out", self.data_width)
+        self._valid_out = self.output("valid_out", 1)
+        self._empty = self.output("empty", 1)
+        self._full = self.output("full", 1)
+
+        # get relevant signals from the storage banks
+        self._data_from_strg = self.input("data_from_strg", self.data_width,
+                                          size=(self.banks,
+                                                self.fw_int),
+                                          explicit_array=True,
+                                          packed=True)
+
+        self._wen_addr = self.var("wen_addr", self.addr_width,
+                                  size=self.banks,
+                                  explicit_array=True,
+                                  packed=True)
+
+        self._ren_addr = self.var("ren_addr", self.addr_width,
+                                  size=self.banks,
+                                  explicit_array=True,
+                                  packed=True)
+
+        self._front_combined = self.var("front_combined", self.data_width,
+                                        size=self.fw_int,
+                                        explicit_array=True,
+                                        packed=True)
+
+        self._data_to_strg = self.output("data_to_strg", self.data_width,
+                                         size=(self.banks,
+                                               self.fw_int),
+                                         explicit_array=True,
+                                         packed=True)
+
+        self._wen_to_strg = self.output("wen_to_strg", self.banks)
+        self._ren_to_strg = self.output("ren_to_strg", self.banks)
+
+        self._num_words_mem = self.var("num_words_mem", 16)  # TODO: Change this width based on memory
```

Comment:
should the 16 be a parameter?

Reply:
Yep. I'm going to do the data width since we will enable chaining (which lets us expand the depth). we will want to revisit this though